### PR TITLE
meson: first try dependency() and fallback find_library()

### LINKED ISSUE #28161: libsystemd doesn't define libcrypt as a dependency
### Component

other

### Is your feature request related to a problem? Please describe

I'm trying to cross-compile `libsystemd` for armv8 using [conan](https://github.com/conan-io/conan) package manager (see related [bug report](https://github.com/conan-io/conan-center-index/issues/15101)).

The problem is that this is failing with:
```
../source/meson.build:991:0: ERROR: C shared or static library 'crypt' not found
```

Which is triggered by [this line](https://github.com/systemd/systemd/blob/main/meson.build#L1064).

I'm not experienced with meson but locally this can be fixed for my case with the following, as conan generates `*.pc` files for the dependencies.
```
diff --git a/meson.build.old b/meson.build
index f38385c..dfa1b55 100644
--- a/meson.build.old
+++ b/meson.build
@@ -988,7 +988,7 @@ threads = dependency('threads')
 librt = cc.find_library('rt')
 libm = cc.find_library('m')
 libdl = cc.find_library('dl')
-libcrypt = cc.find_library('crypt')
+libcrypt = dependency('crypt')
 
 crypt_header = conf.get('HAVE_CRYPT_H') == 1 ? '''#include <crypt.h>''' : '''#include <unistd.h>'''
 foreach ident : [
```

Is this a valid fix that can be upstreamed to systemd?

### Describe the solution you'd like

See the given patch, so that libcrypt can be found outside the root-file system.

### Describe alternatives you've considered

_No response_

### The systemd version you checked that didn't have the feature you are asking for

251.4

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #28161.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -1061,7 +1061,8 @@ threads = dependency('threads')
 librt = cc.find_library('rt')
 libm = cc.find_library('m')
 libdl = cc.find_library('dl')
-libcrypt = cc.find_library('crypt')
+libcrypt = dependency('libcrypt')
+libcap = dependency('libcap')
```
Comment (bluca) [inline meson.build:1064]: fallbacks are missing now?

Comment (yuwata) [reply meson.build:1064]: Not necessary anymore, as all supported releases have .pc for the library.

Comment (bluca) [reply meson.build:1064]: please mention that in the commit message (ideally with the versions that introduced the .pc files) as it wasn't clear, it just says `meson: first try dependency(), then fallback to find_library()`

Comment (bluca) [reply meson.build:1064]: what about libcrypt and libacl?

Comment (yuwata) [reply meson.build:1064]: Oops. The same. Will update the comment.

File: meson.build [hunk 2]
```diff
@@ -1269,7 +1264,7 @@ conf.set10('ENABLE_POLKIT', install_polkit)
 
 want_acl = get_option('acl')
 if want_acl != 'false' and not skip_deps
-        libacl = cc.find_library('acl', required : want_acl == 'true')
+        libacl = dependency('libacl', required : want_acl == 'true')
```
Comment (bluca) [inline meson.build:1268]: this is also not mentioned in the commit message

File: meson.build [hunk 3]
```diff
@@ -1456,8 +1458,12 @@ conf.set10('HAVE_QRENCODE', have)
 
 want_gcrypt = get_option('gcrypt')
 if want_gcrypt != 'false' and not skip_deps
-        libgcrypt = cc.find_library('gcrypt', required : want_gcrypt == 'true')
-        libgpg_error = cc.find_library('gpg-error', required : want_gcrypt == 'true')
+        libgcrypt = dependency('libgcrypt', required : want_gcrypt == 'true')
```
Comment (bluca) [inline meson.build:1462]: and this as well - are they really all supported from the exact same distribution version?

Comment (yuwata) [reply meson.build:1462]: Yes. All libraries are checked in Debian and Ubuntu package search page.

**Review by bluca — APPROVED** [review]

File: meson.build [hunk 4]
```diff
@@ -1568,8 +1574,11 @@ conf.set10('HAVE_ZLIB', have)
 
 want_bzip2 = get_option('bzip2')
 if want_bzip2 != 'false' and not skip_deps
-        libbzip2 = cc.find_library('bz2',
-                                   required : want_bzip2 == 'true')
+        libbzip2 = dependency('bzip2', required : false)
+        if not libbzip2.found()
+                # Debian and Ubuntu do not provide the .pc file.
+                libbzip2 = cc.find_library('bz2', required : want_bzip2 == 'true')
+        endif
```
Comment (eli-schwartz) [inline meson.build:1581]: There is no .pc file for Debian/Ubuntu to not-provide. bzip2 doesn't have one and never has. Although I did submit a patch to add one: https://sourceware.org/pipermail/bzip2-devel/2022q2/000169.html

Comment (yuwata) [reply meson.build:1581]: Interesting. Maybe, fedora picks your patch.

**Pull request merged**